### PR TITLE
feat: initial Druid dialect

### PR DESCRIPTION
**Pull request by betodealmeida**
We're using `sqlglot` to parse SQL in Apache Superset. The generic dialect has worked fine so far for parsing Druid queries, but we noticed that the [FLOOR (date and time)](https://druid.apache.org/docs/latest/querying/sql-functions#floor-date-and-time) and [CEIL (date and time)](https://druid.apache.org/docs/latest/querying/sql-functions#ceil-date-and-time) functions cannot be parsed:

```python
>>> sqlglot.parse_one("SELECT FLOOR(__time TO WEEK) FROM t")
...
sqlglot.errors.ParseError: Expecting ). Line 1, Col: 22.
  SELECT FLOOR(__time TO WEEK) FROM t
```

This PR introduces an initial dialect for Druid that handles both `FLOOR` and `CEIL` when they are used for time rounding.

### CODE REVIEW

File: sqlglot/dialects/druid.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+from sqlglot import exp, generator, parser
+from sqlglot.dialects.dialect import Dialect
+from sqlglot.tokens import TokenType
+
+
+class TemporalFloor(exp.Func):
```
Comment (tobymao) [inline sqlglot/dialects/druid.py:6]: we don't have dialect level expressions, please move these to expressions.py, you also need to add rendering in the generic dialect, we need to ensure all sql is convertible to generic.

File: tests/dialects/test_druid.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class TestDruid(Validator):
+    dialect = "druid"
+
+    def test_druid(self):
+        self.validate_identity("SELECT CEIL(__time TO WEEK) FROM t")
+        self.validate_identity("SELECT CEIL(col) FROM t")
+        self.validate_identity("SELECT FLOOR(__time TO WEEK) FROM t")
+        self.validate_identity("SELECT FLOOR(col) FROM t")
```
Comment (tobymao) [inline tests/dialects/test_druid.py:11]: add a test

validate_all(
 floor(x to week),
write: {
  "": ...
}
)

to make sure transpilation can happen to other dialects, they don't need to be correct but they shouldn't crash

File: sqlglot/dialects/druid.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            exp.DataType.Type.INT: "LONG",
+            exp.DataType.Type.FLOAT: "FLOAT",
+            exp.DataType.Type.DOUBLE: "DOUBLE",
+            exp.DataType.Type.TEXT: "STRING",
+            exp.DataType.Type.ARRAY: "ARRAY",
+        }
+
+        TRANSFORMS = {
+            **generator.Generator.TRANSFORMS,
+            TemporalFloor: lambda self, e: f"FLOOR({e.this} TO {e.unit})",
```
Comment (tobymao) [inline sqlglot/dialects/druid.py:103]: this should go in generic generator or else it's gonna crash

File: sqlglot/dialects/druid.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    arg_types = {"unit": True, "this": True}
+
+    @property
+    def unit(self) -> exp.Expression:
+        return self.args["unit"]
+
+
+class Druid(Dialect):
+    class Parser(parser.Parser):
+        def __init__(self, *args, **kwargs):
```
Comment (tobymao) [inline sqlglot/dialects/druid.py:36]: delete this section

**Comment by georgesittas** [comment]
Closing this due to inactivity, feel free to open another one if/when you address Toby's comments.

**Comment by betodealmeida** [comment]
@georgesittas yeah, just haven't had the time. I'll reopen when it's ready, thanks!

**Pull request closed without merging**